crates/sui-graphql-rpc/src/types/cursor.rs

```diff
@@ -0,0 +1,357 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{fmt, ops::Deref};
+
+use async_graphql::{
+    connection::{CursorType, OpaqueCursor},
+    *,
+};
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::{config::ServiceConfig, error::Error};
+
+/// Wrap the `OpaqueCursor` type to implement Scalar for it.
+pub(crate) struct Cursor<C>(OpaqueCursor<C>);
+
+/// Connection field parameters parsed into a single type that encodes
+#[derive(Debug)]
+pub(crate) struct Page<C> {
+    after: Option<Cursor<C>>,
+    before: Option<Cursor<C>>,
+    limit: u64,
+    side: Side,
+}
+
+/// Whether the page is extracted from the beginning or the end of the range bounded by the cursors.
+#[derive(PartialEq, Eq, Debug)]
+enum Side {
+    Prefix,
+    Suffix,
+}
+
+impl<C> Cursor<C> {
+    pub(crate) fn new(cursor: C) -> Self {
+        Cursor(OpaqueCursor(cursor))
+    }
+}
+
+impl<C> Page<C> {
+    pub(crate) fn from_params(
+        config: &ServiceConfig,
+        first: Option<u64>,
+        after: Option<Cursor<C>>,
+        last: Option<u64>,
+        before: Option<Cursor<C>>,
+    ) -> Result<Self> {
+        let limits = &config.limits;
+        let page = match (first, after, last, before) {
+            (Some(_), _, Some(_), _) => return Err(Error::CursorNoFirstLast.extend()),
+
+            (limit, after, None, before) => Page {
```

Comment:
Suggestion: a comment like defaults to side::prefix with default page limit if neither last nor first are provided

Reply:
Good suggestion!  Added this as a doc comment on the function.